monai/utils/module.py

```diff
@@ -193,7 +211,127 @@ def load_submodules(basemod, load_all: bool = True, exclude_pattern: str = "(.*[
     return submodules, err_mod
 
 
+class ComponentScanner:
+    """
+    Scan all the available classes and functions in the specified packages and modules.
+    Map the all the names and the module names in a table.
+
+    Args:
+        pkgs: the expected packages to scan modules and parse component names in the config.
+        modules: the expected modules in the packages to scan for all the components.
+            for example, to parser "LoadImage" in config, `pkgs` can be ["monai"], `modules` can be ["transforms"].
+        excludes: if any string of the `excludes` exists in the full module name, don't import this module.
+
+    """
+
+    def __init__(
+        self, pkgs: Sequence[str], modules: Optional[Sequence[str]] = None, excludes: Optional[Sequence[str]] = None
+    ):
+        self.pkgs = pkgs
+        self.modules = [] if modules is None else modules
+        self.excludes = [] if excludes is None else excludes
+        self._components_table = self._create_table()
+
+    def _create_table(self):
+        table: Dict[str, List] = {}
+        for pkg in self.pkgs:
+            package = import_module(pkg)
+
+            for _, modname, _ in walk_packages(path=package.__path__, prefix=package.__name__ + "."):
+                # if no modules specified, load all modules in the package
+                if all(s not in modname for s in self.excludes) and (
+                    len(self.modules) == 0 or any(name in modname for name in self.modules)
+                ):
+                    try:
+                        module = import_module(modname)
+                        for name, obj in inspect.getmembers(module):
+                            if (inspect.isclass(obj) or inspect.isfunction(obj)) and obj.__module__ == modname:
+                                if name not in table:
+                                    table[name] = []
+                                table[name].append(modname)
+                    except ModuleNotFoundError:
+                        pass
+        return table
+
+    def get_component_module_name(self, name):
+        """
+        Get the full module name of the class / function with specified name.
+        If target component name exists in multiple packages or modules, return all the paths.z
+
+        Args:
+            name: name of the expected class or function.
+
+        """
+        mods = self._components_table.get(name, None)
+        if isinstance(mods, list) and len(mods) == 1:
+            mods = mods[0]
+        return mods
+
+
+def locate(path: str):
+    """
+    Locate an object by name or dotted path, importing as necessary.
+    Refer to Hydra: https://github.com/facebookresearch/hydra/blob/v1.1.1/hydra/_internal/utils.py#L554.
+
+    Args:
+        path: full path of the expected component to locate and import.
+
+    """
+    parts = [part for part in path.split(".") if part]
+    for n in reversed(range(1, len(parts) + 1)):
```

Comment:
This might be a little better:

```python
parts = [p for p in path.split(".") if p]
for n in range(len(parts), 0, -1)):
        try:
            obj = import_module(".".join(parts[:n]))
            break
        except Exception as exc_import:
            if n == 1:
                raise ImportError(f"Cannot import module '{path}'.") from exc_import
```

Reply:
Sure, updated in the latest commit.

Thanks.